### PR TITLE
Add EthernetPacket field to PacketOut

### PR DESCRIPTION
**Pull request by GraysonWu**
Provide an easy way to send an ethernet packet via PacketOut without filling all L2/3/4 headers.

### CODE REVIEW

**Comment by wenyingd** [comment]
I didn't get your purpose of this patch, do you mean you just want to get the eth header in a packetOut message?

**Comment by GraysonWu** [comment]
Yes, in some situations, for example, we received a packetIn and we only need the info in this packetIn to execute specific process. After the process, we need to forward this packet via packetOut without any modifications. So IMO adding this patch can make this easier. Otherwise, we need to retrieve and format all L2/3/4 headers.

File: ofctrl/ofPacket.go [hunk 1]
```diff
@@ -25,6 +25,8 @@ type PacketOut struct {
 	ICMPHeader *protocol.ICMP
 
 	ARPHeader *protocol.ARP
+
+	EthernetPacket *util.Message
```
[the remark this answers is not in the document]
Comment (GraysonWu) [reply ofctrl/ofPacket.go]: Done.

**Review by wenyingd — APPROVED** [review]
LGTM

File: ofctrl/ofPacket.go [hunk 2]
```diff
@@ -25,6 +25,14 @@ type PacketOut struct {
 	ICMPHeader *protocol.ICMP
 
 	ARPHeader *protocol.ARP
+
+	// EthernetPacket provide a convenient way to fill an ethernet packet into
```
Comment (hongliangl) [inline ofctrl/ofPacket.go:29]:
```suggestion
	// EthernetPacket provides a convenient way to generate an Ethernet packet into
```

File: ofctrl/ofPacket.go [hunk 3]
```diff
@@ -25,6 +25,14 @@ type PacketOut struct {
...
+	// packetOut without manually setting all L2/3/4 header and data.
```
Comment (hongliangl) [inline ofctrl/ofPacket.go:30]:
```suggestion
	// packetOut without manually filling all L2/3/4 headers and fields.
```

File: ofctrl/ofPacket.go [hunk 4]
```diff
@@ -25,6 +25,14 @@ type PacketOut struct {
...
+	// If EthernetPacket is provided, all L2/3/4 related fields, namely SrcMAC, DstMAC,
+	// IPHeader, IPv6Header, TCPHeader, UDPHeader, ICMPHeader and ARPHeader, will be
+	// ignored. If EthernetPacket is not used (is nil), an ethernet packet will be
```
Comment (hongliangl) [inline ofctrl/ofPacket.go:33]:
```suggestion
	// ignored. If EthernetPacket is not used (is nil), an Ethernet packet will be
```

Comment (hongliangl) [reply ofctrl/ofPacket.go:33]: Maybe rephrase this comments. I'm a little confused about 
```
// If EthernetPacket is provided, all L2/3/4 related fields, namely SrcMAC, DstMAC,
	// IPHeader, IPv6Header, TCPHeader, UDPHeader, ICMPHeader and ARPHeader, will be
	// ignored. If EthernetPacket is not used (is nil), an ethernet packet will be
// generated based on those L2/3/4 related fields.
```

I thought that you wanted to say that: when calling `getEthernetPacket`, if field `EthernetPacket` has been generated, return it directly; otherwise, generate and return it.

File: ofctrl/ofPacket.go (hunk 2 above)
Comment (GraysonWu) [reply ofctrl/ofPacket.go]: Done.

File: ofctrl/ofPacket.go (hunk 3 above)
Comment (GraysonWu) [reply ofctrl/ofPacket.go]: Done.

File: ofctrl/ofPacket.go (hunk 4 above)
Comment (GraysonWu) [reply ofctrl/ofPacket.go]: Rephrased. And move some comments to `getEthernetPacket`. Hope you can take another look.

**Review by hongliangl — APPROVED** [review]
LGTM, could you help double check @wenyingd

**Review by wenyingd — APPROVED** [review]
LGTM
